### PR TITLE
Advance index

### PR DESCRIPTION
**Pull request by JackCaoG**
This is to fix https://github.com/pytorch/xla/issues/3184. After the pytorch/pytorch#66506, index and base of an index can be on different device and pt/xla needs to handle this case by moving all tensors to the xla device.

FYI: @zou3519

### CODE REVIEW

File: torch_xla/csrc/aten_xla_type.cpp
Comment (miladm) [inline torch_xla/csrc/aten_xla_type.cpp:1654]: Why don't we apply this `if/else` logic in `GetCAnonicalIndexInfo()`? Is that because we need `device` for `bridge::GetOrCreateXlaTensor(values, *device)`? If yes, the next question is, why do we believe an upstream `value` tensor "can" be on a different device?

(My understanding is `base` and `index` are the root cause of this failure; correct me if I am wrong.)

File: torch_xla/csrc/aten_xla_type.cpp
Comment (miladm) [inline torch_xla/csrc/aten_xla_type.cpp:1600]: Can we test this scenario? (i.e. having `base` and `index` on diff devices)

**Review by miladm — CHANGES_REQUESTED** [review]
Thanks @JackCaoG. I put in a few comments. Let me know what you think.

File: torch_patches/.torch_pin
Comment (JackCaoG) [reply torch_patches/.torch_pin:1]: https://github.com/pytorch/xla/issues/3184 has a good summary.

File: torch_xla/csrc/aten_xla_type.cpp
Comment (JackCaoG) [reply torch_xla/csrc/aten_xla_type.cpp:1600]: Pytorch side has a test for this use case, https://github.com/pytorch/pytorch/pull/66506/files

File: torch_xla/csrc/aten_xla_type.cpp
Comment (JackCaoG) [reply torch_xla/csrc/aten_xla_type.cpp:1654]: I put it here mostly because we also handle the binary op tensor [device mismatch](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/aten_xla_type.cpp#L115) in this file. 

Index is of type `c10::List<c10::optional<at::Tensor>>` which is something `GetOrCreateXlaTensor` can't handle. We generally not trying to change the `at::tensor` device type but create the `XLATensor` on the correct device. `at::tensor` -> `XLATensor` happens in this file.

Comment (miladm) [reply torch_xla/csrc/aten_xla_type.cpp:1654]: Got it. Given it's used in three spots, why don't we do this logic as a separate function call?

Comment (JackCaoG) [reply torch_xla/csrc/aten_xla_type.cpp:1654]: It is a bit awkward to add a helper function like

```
bridge::GetXlaDevice(at::Tensor tensor, vector<at::Tensor> tensors);
```
one cleaner way of doing this is to append the single tensor to the end of the vector and use the vector version of the `GetXlaDevice`, but that felt like an overkill since we either need to create a new vector or manually pop the `base` after this call.

Comment (miladm) [reply torch_xla/csrc/aten_xla_type.cpp:1654]: Yeah, maybe this implementation is the most appropriate of all.

**Review by miladm — APPROVED** [review]
LGTM - Thanks @JackCaoG

**Comment by JackCaoG** [comment]
@zou3519 Yup, I will merge this one once pytorch one merged

**Pull request merged**
